pkg/controllers/utils/termination.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"context"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"knative.dev/pkg/apis"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+)
+
+// EnsureTerminated is a helper function that takes a v1beta1.NodeClaim and calls cloudProvider.Delete() if status condition
+// on nodeClaim is not terminating. If it is terminating then it will check if the kubelet process is still running. It will
+// return an error and a boolean that indicates if the reconciliation should be re-enqueued.
+func EnsureTerminated(ctx context.Context, c client.Client, nodeClaim *v1beta1.NodeClaim, cloudProvider cloudprovider.CloudProvider) (bool, error) {
+	// Check if the status condition on nodeClaim is Terminating
+	if !nodeClaim.StatusConditions().GetCondition(v1beta1.Terminating).IsTrue() {
+		// If not then call Delete on cloudProvider to trigger termination and always requeue reconciliation
+		err := cloudProvider.Delete(ctx, nodeClaim)
+		if cloudprovider.IgnoreNodeClaimNotFoundError(err) != nil {
+			return false, fmt.Errorf("terminating cloudprovider instance, %w", err)
+		}
+		nodeClaim.StatusConditions().SetCondition(apis.Condition{
+			Type:     v1beta1.Terminating,
+			Status:   v1.ConditionTrue,
+			Severity: apis.ConditionSeverityInfo,
+		})
+		// We call Update() here rather than Patch() because patching a list with a JSON merge patch
+		// can cause races due to the fact that it fully replaces the list on a change
+		// https://github.com/kubernetes/kubernetes/issues/111643#issuecomment-2016489732
+		if updateErr := c.Status().Update(ctx, nodeClaim); updateErr != nil {
+			if apierrors.IsConflict(updateErr) {
+				return true, nil
+			}
+			return false, client.IgnoreNotFound(updateErr)
+		}
+		if cloudprovider.IsNodeClaimNotFoundError(err) {
```

Comment:
This error checking below is a tad awkward. Is there a way to make the error scoped closer to where it's actually generated so that we don't have to create separate error variables. Creating separate error variables should be done sparingly because it's really easy to mix up where the error is coming from and how it should be handled if it's far away from the thing that generated the call